### ISSUE TITLE
Feature request: Add Config(const char * filename) and Config(FILE * stream) constructors

### ISSUE THREAD

**Issue by Shulyaka** [hyperrealm/libconfig#59]
Hi,

It would be convenient to have a constructor that reads a file automatically. Reading a config file is usually an operation that is executed only once and prevents doing other things with Config until it's done.
It would also allow the following code:

class MyClass
{
  Config cfg;
  const int myparam;
  MyClass(const char *configfile) : cfg(configfile), myparam(cfg.lookup("myparam")) {};
}

Best regards,
Denis Shulyaka

**Comment by hyperrealm** [comment]
It's generally not a good idea for C++ constructors to do complicated tasks.

**Comment by jltallon** [comment]
Shuiyaka, You can definitively do that yourself if you want to.
I wouldn't use an initializer list for myparam, but assign it within the constructor's body, though.

Mark, up to you to label it as wontfix.

**Comment by hyperrealm** [comment]
I'm against this because tasks that are potentially long-running or that can throw exceptions should not be in constructors. If, for example, you have a global variable that initialized a Config with a filename parameter, and the parsing failed and threw an exception, then the exception would occur during static initialization before main() even started executing. These types of things can be hard to debug. So the current API is intentionally inconvenient, to encourage the developer to think about error handling and such when reading/parsing a configuration.

**Comment by Shulyaka** [comment]
Ok. Well, I accept your decision, but to be honest I disagree that it would
be hard to debug such config parsing errors, the program would just
terminate and print the exception, which is enough. I also personally don't
see any evil in complex constructors.
jltallon, if i assign myparam in constructor I cannot make it const.

However it is entirely your decision. After all, I can always make a
wrapper class to make all my dirty stuff.

Thank you!
12 февр. 2016 г. 1:37 пользователь "hyperrealm" <notifications@github.com>
написал:
